### PR TITLE
Added program for parenthesis matching

### PR DESCRIPTION
**Pull request by kanak22**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

YES

### Description
I have added a program for parenthesis matching in C

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
#3446

### CODE REVIEW

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@kanak22 Update the readme file.

**Comment by kanak22** [comment]
> @kanak22 Update the readme file.

I did. Kindly check :)

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Few changes: 

- Add Multi-Line Comments 
- Add a beginner-friendly description at the top
- Use proper code Format 
- Remove extra lines of code and spaces
- Add sample I/O at the end of the Code
- Add Time and Space Complexity at the end of the Code   
- Add new line at the EOF

**Comment by kanak22** [comment]
> Few changes:
> 
> * Add Multi-Line Comments
> * Add a beginner-friendly description at the top
> * Use proper code Format
> * Remove extra lines of code and spaces
> * Add sample I/O at the end of the Code
> * Add Time and Space Complexity at the end of the Code
> * Add new line at the EOF

sure.

**Comment by kanak22** [comment]
I did the asked changes still if there is anything you would like to be changed kindly let me know :)

File: C/stack/Parenthesis_Matching.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+Valid expression
+
+Press any key to continue . . .
+
+    2.
+Enter an algebraic expression : [a(b+c}]
+Mismatched parentheses are : ( and }
+Invalid expression
+
+Press any key to continue . . .
```
Comment (iamrajiv) [inline C/stack/Parenthesis_Matching.c]: @kanak22 Remove `Press any key to continue . . .` and also remove unnecessary line breaks.

**Review by iamrajiv — CHANGES_REQUESTED** [review]

File: C/stack/Parenthesis_Matching.c (hunk 1 above)
Comment (kanak22) [reply C/stack/Parenthesis_Matching.c]: I removed Press any key to continue . . . 
but can you suggest me where there are unnecessary line breaks :)

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@kanak22 Format your code like this

```c
/* This program is of parenthesis matching which is an application of stack in C */
/* TIME COMPLEXITY: O(n) */

#include<stdio.h>
#include<string.h>
#include<stdlib.h>

//global declarations
#define MAX 30
int top = -1;
int stack[MAX];

//function prototypes
void push(char);
char pop();
int match(char a, char b);
int check(char []);

//main
int main() {
    //declaration
    char exp[MAX];
    int valid;

    //taking input
    printf("Enter an algebraic expression : ");
    gets(exp);

    //calling function check
    valid = check(exp);

    //checking if valid is equal to 1 or not
    if (valid == 1)
    {  printf("Valid expression\n");}
    else
    {printf("Invalid expression\n");}

    return 0;

}

//function to check parenthesis
int check(char exp[] ) {
    //declaration
    int i, n;
    char temp;
    n = strlen(exp);
    /*loop for traversing through the expression
    storing(pushing) parenthesis into stack
    poping if the match of that parenthesis is found*/
    for (i = 0; i < n; i++) {
        if (exp[i] == '(' || exp[i] == '{' || exp[i] == '[')
            push(exp[i]);
        if (exp[i] == ')' || exp[i] == '}' || exp[i] == ']') {
            //stack empty
            if (top == -1) {
                //when there are more parenthesis on the right
                printf("Right parentheses are more than left parentheses\n");
                return 0;
            } else {
                temp = pop();
                //calling function match to match parenthesis
                if (!match(temp, exp[i])) {
                    //to print the parenthesis  which gets mismatched
                    printf("Mismatched parentheses are : ");
                    printf("%c and %c\n", temp, exp[i]);
                    return 0;
                }
            }
        }
    }

    //if stack is empty
    if (top == -1) {
        //if stack gets empty
        printf("Balanced Parentheses\n");
        return 1;
    } else {
        //if stack isn't empty
        printf("Left parentheses more than right parentheses\n");
        return 0;
    }
}

//function to match parenthesis
int match(char a, char b) {
    //if a=[ and b=]
    if (a == '[' && b == ']')
    {  return 1;}
    //if a={ and b=}
    if (a == '{' && b == '}')
    {  return 1;}
    //if a=( and b=)
    if (a == '(' && b == ')')
    {  return 1;}
    return 0;
}

//function to push element into stack
void push(char item) {
    if (top == (MAX - 1)) {
        //printing message if stack is full
        printf("Stack Overflow\n");
        return;
    }
    top = top + 1;
    stack[top] = item;
}

//function to pop element from stack
char pop() {
    if (top == -1) {
        //printing message if stack is empty
        printf("Stack Underflow\n");
        exit(1);
    }
    return (stack[top--]);
}

/* sample input/output */
/*  1.
Enter an algebraic expression : (((5+7)*6)/2)
Balanced Parentheses
Valid expression
    2.
Enter an algebraic expression : [a(b+c}]
Mismatched parentheses are : ( and }
Invalid expression

*/
```

**Comment by kanak22** [comment]
Oh okay ...got it ...doing it :)

**Comment by kanak22** [comment]
I did this also. :)

**Review by ankitaggarwal23 — COMMENTED** [review]
Code plagiarized from [here](https://www.codezclub.com/c-check-balanced-parentheses-using-stack/).

**Pull request closed without merging**